### PR TITLE
Introduce a cluster-aware addressing scheme

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
Please see commit messages for more details.

```release-note
Introduce a cluster-aware addressing scheme
```

### CODE REVIEW

File: pkg/k8s/endpoints.go [hunk 1]
```diff
@@ -119,10 +120,10 @@ func ParseEndpoints(ep *slim_corev1.Endpoints) (ServiceID, *Endpoints) {
 
 	for _, sub := range ep.Subsets {
 		for _, addr := range sub.Addresses {
-			backend, ok := endpoints.Backends[addr.IP]
+			backend, ok := endpoints.Backends[cmtypes.NewIPCluster(addr.IP, 0)]
```
Comment (tommyp1ckles) [inline pkg/k8s/endpoints.go:123]: Is everything in k8s/endpoints always going to refer to the local cluster?

Comment (YutaroHayakawa) [reply pkg/k8s/endpoints.go:123]: Yes, ClusterID 0 is treated as a local cluster.

**Comment by YutaroHayakawa** [comment]
Cilium-PR-K8s-1.16-kernel-4.9 is a known flake: https://github.com/cilium/cilium/issues/20288#issuecomment-1212653930

**Comment by YutaroHayakawa** [comment]
/test-1.23-4.19

**Review by tommyp1ckles — APPROVED** [review]

**Comment by YutaroHayakawa** [comment]
Tests are passing except `test-1.16-4.9`, which hits a known flake. Making this ready to merge.

**Comment by YutaroHayakawa** [comment]
Rebased with the latest master

**Comment by YutaroHayakawa** [comment]
I noticed I need one more review for merging. Removing ready-to-merge.

**Review by ldelossa — APPROVED** [review]
LGTM

**Comment by YutaroHayakawa** [comment]
Fixed conflict with latest master

**Comment by YutaroHayakawa** [comment]
Reviews are in and ci-external-workloads is failing with timeout, but connectivity tests are actually passing (https://github.com/cilium/cilium/issues/21143) making this ready to merge.

**Comment by YutaroHayakawa** [comment]
Sorry, let me remove the ready-to-merge again. We may want to change `IPCluster.ip` field from string to `netip.Addr`.

**Comment by YutaroHayakawa** [comment]
Yes, once I confirm that changing string -> netip.Addr doesn't have any problem, I'll replace this PR with it.

**Comment by YutaroHayakawa** [comment]
Reworked in https://github.com/cilium/cilium/pull/21161. Let me close this PR.

**Pull request closed without merging**